File: Csharp/CodeWars/src/Resolutions/CatYearsDogYears.cs
namespace CodeWars.Resolutions;

public class CatYearsDogYears
{
    public static int[] HumanYearsCatYearsDogYears(int humanYears)
    {
        return [CalculateYears(humanYears, "human"),
                CalculateYears(humanYears, "cat"),
                CalculateYears(humanYears, "dog")];
    }

    private static int CalculateYears(int humanYears, string species)
    {
        if (species.Equals("human")) return humanYears;

        int multiplicator = species.Equals("cat") ? 4 : 5;
        int[] initialAge = [15, 9];

        int[] ages = [.. initialAge.Concat(Enumerable.Repeat(multiplicator, humanYears))];

        return ages.Take(humanYears).Sum();
    }
}

File: TypeScript/Codewars/src/resolutions/Cat-years-Dog-years.ts
const TypeYears = {
	Cat: 1,
	Dog: 2,
};

export function humanYearsCatYearsDogYears(humanYears: number): [number, number, number] {
	const catYears = calculateYears(humanYears, TypeYears.Cat);
	const dogYears = calculateYears(humanYears, TypeYears.Dog);

	return [humanYears, catYears, dogYears];
}

function calculateYears(years: number, Tp: number): number {
	const yearsMultiplications = Tp === TypeYears.Cat ? 4 : 5;
	const array: number[] = [];

	array.push(...[15, 9], ...new Array(years).fill(yearsMultiplications));

	return array.slice(0, years).reduce((prev, curr) => prev + curr, 0);
}

console.log(humanYearsCatYearsDogYears(3));
